signing: Add signctl tool and documentation. (#646)
ARMmbed/meta-mbl: openembedded-core-mbl/scripts/signctl/README.md

# Mbed Linux OS RPi3 Signing Tool

## Overview

> The `signctl` tool automates the basic signing flow for Raspberry PI 3. For more information on the background and details of the signing operations see [the basic signing flow document](https://github.com/ARMmbed/meta-mbl/blob/warrior-dev/docs/basic-signing-flow.md).

Author: Added relative paths and indirection in bd16b66

> The signing tool uses [Hashicorp Vault](https://www.hashicorp.com/products/vault/) to generate and store the certificates required to verify the ATF boot chain.

Reviewer: GLOBAL: We should refer to Trusted Firmware as TF-A.

The Vault secrets model uses 'paths' to 'mount' secrets engines. The 'path' is a REST API endpoint and can be accessed via HTTP requests.

For more background on the Vault secrets engine model see [this link](https://www.vaultproject.io/intro/getting-started/secrets-engines#secrets-engines)

In the Vault PKI engine, each 'mount' of the secrets engine is usually a root CA or intermediate CA, which can then sign and issue certificates from the endpoints created under the base path. Only one CA is allowed per instance of the PKI engine.

The `signctl` tool uses a CA configured at its own mount to represent the private key for each node in the certificate chain of trust.
The tool then requests leaf certificates from the appropriate mount path, so the corresponding cert is signed by the correct private key.

> For more background on the PKI secrets engine see [this link](https://www.vaultproject.io/docs/secrets/pki/index.html#one-ca-certificate-one-secrets-engine).

Reviewer: Re-word like above to remove "this link" please.

Author: Re-worded in bd16b66

## Install Prerequisites

> Before installing the `signctl` tool, you should set up a local dev instance of Vault. For more background see [this link](https://www.vaultproject.io/docs/install/#compiling-from-source).

Reviewer: ```suggestion
Before installing the `signctl` tool, you should set up a local dev instance of Vault. The [Vault website](https://www.vaultproject.io/docs/install/#compiling-from-source) provides background information on this.
```

    write       Write data, configuration, and secrets
    delete      Delete secrets and configuration
    list        List data or secrets
    login       Authenticate locally
    agent       Start a Vault agent
    server      Start a Vault server
    status      Print seal and HA status
    unwrap      Unwrap a wrapped secret

Other commands:
    audit          Interact with audit devices
    auth           Interact with auth methods
    kv             Interact with Vault's Key-Value storage
    lease          Interact with leases
    namespace      Interact with namespaces
    operator       Perform operator-specific tasks
    path-help      Retrieve API help for paths
    plugin         Interact with Vault plugins and catalog
    policy         Interact with policies
    print          Prints runtime configurations
    secrets        Interact with secrets engines
    ssh            Initiate an SSH session
    token          Interact with tokens

```

## Installing the Signing Tool

Ubuntu 16.04: You must install some packages before following the installation procedure.

```
$ apt-get install git make u-boot-tools build-essential libssl-dev libffi-dev
```

> First you should create a Python virtual environment. If you don't understand what a Python virtual environment is see: https://docs.python.org/3/tutorial/venv.html

Reviewer: ```suggestion
First you should create a Python virtual environment following the steps below. This [tutorial](https://docs.python.org/3/tutorial/venv.html) explains more about Python virtual environments.
```

<span class="notes">**Note:** If you don't want to create a virtual environment, you can possibly install with `pip --user`. Don't ever install packages into the Ubuntu "system" Python `site-packages`.

> Installing outside a virtual environment is not recommended. You're on your own if you elect to use another installation method.</span>

Reviewer: ```suggestion
Installing outside a virtual environment is not recommended, tested or supported.</span>
```

> Follow the steps below:

Reviewer: ```suggestion
```

optional arguments:
  -h, --help            show this help message and exit
  --backend-url BACKEND_URL
```
```
usage: signctl.py generate [-h] [--cot-keys] [--hab-srks]
                           [--output-dir OUTPUT_DIR]

optional arguments:
  -h, --help            show this help message and exit
  --cot-keys            Generate an ATF chain of trust keychain.
  --hab-srks            Generate super root public keys for NXP HAB.
  --output-dir OUTPUT_DIR
                        The output directory for the HAB SRKs (public part).
```
```
usage: signctl.py sign-atf [-h] [--rpi-armstub8 RPI_ARMSTUB8] --fip FIP
                           [FIP ...] [--patch-rotkey]
                           [--output-dir OUTPUT_DIR]

optional arguments:
  -h, --help            show this help message and exit
  --rpi-armstub8 RPI_ARMSTUB8
                        Path to armstub8.bin
  --fip FIP [FIP ...]   Path to one or more FIPs to be signed.
  --patch-rotkey        Patch root of trust public key hash in to BL1 and BL2.
  --output-dir OUTPUT_DIR
                        Output directory for signed images and the original
                        fip components.
```

> ## Generating Chain of Trust Keychain

Reviewer: ```suggestion
## Generating Chain of Trust (CoT) Keychain
```

You can generate the keychain used for the ATF CoT using the `signctl generate --cot-keys` command.

Currently the `generate --cot-keys` command creates all of the required CAs up to the root of trust. The private keys are stored in the Vault.

In the release version, the `signctl` tool will support generating individual keypairs in the keychain, allowing for key revocation and access restriction to high security keys.

> <span class="notes">**Note:** The tool only supports signing for a single device at the moment. The release version will support multiple devices; we will provide documentation and guidance for setting access control policies in Vault, and for locking down the Vault machine for production use. **The current flow is only for development use.**</span>

Reviewer: Could add this as a point in the overview section on restrictions above.

Author: Moved to the overview section in bd16b66

3. Run the following command
```
$ signctl generate --cot-keys
```

<span class="notes">**Note:** If using the `--backend-url` flag, it should be entered before the command. For example:

```
$ signctl --backend-url 127.0.0.1:8200 generate --cot-keys
```
</span>

The keys should be shown as generated in the Vault server output. If the `signctl` tool returns with no output it means the keys were successfully generated.

## Signing RPi3 Boot Components

After the keys have been generated and stored in Vault, you can use them to sign the certificates needed to verify the FIP.

To produce the signed leaf certificates and pack them into a FIP use the `signctl sign-atf` command.

If you saved your Vault server URL in a `.signctl.conf` file as described in the previous section, the `signctl` tool will automatically connect to the Vault server. If not you will have to pass the server URL again on the command line.

> You can also patch the root of trust public keys into BL1 and BL2 by using the `--patch-rotkeys` command. The background for this is described in the [the basic signing flow document](https://github.com/ARMmbed/meta-mbl/blob/warrior-dev/docs/basic-signing-flow.md).

Reviewer: As above use indirect links and relative links for docs.

Author: Added indirect links in bd16b66

│   ├── nt-fw-key-cert.bin
│   ├── soc-fw-cert.bin
│   ├── soc-fw-key-cert.bin
│   ├── tb-fw-cert.bin
│   ├── tos-fw-cert.bin
│   ├── tos-fw-key-cert.bin
│   └── trusted-key-cert.bin
└── signed_images
    ├── armstub8_new.bin
    ├── fip1_new.bin
    └── fip2.bin
```

The signed_images directory contains the signed `armstub8.bin` and `fip2.bin`.

## Writing the Signed Components to the SD Card
Connect your SD card, containing an Mbed Linux OS Raspberry PI Image, to your pc. Ensure the partitions are unmounted.

Write `fip2.bin` to a "raw partition" on the flash.
Assuming your SD card is located at `/dev/sdc` run the following command.

```
$ sudo dd if=signed_images/fip2.bin of=/dev/sdc bs=512 seek=2048
1650+1 records in
1650+1 records out
845177 bytes (845 kB, 825 KiB) copied, 0.148319 s, 5.7 MB/s
```

Mount the boot partition and copy `armstub8_new.bin` to it.

```
$ udisksctl mount -b /dev/sdc1
Mounted /dev/sdc1 at /media/<user>/boot.
```
```
$ cp signed_imgs/armstub8_new.bin /media/<user>/boot/armstub8.bin
```
```
$ sync
```
```
$ udisksctl unmount -b /dev/sdc1
Unmounted /dev/sdc1.
```

You can now place the SD card in your RPi3 and it will verify and load the FIP image.

> If you're using an Mbed Linux image it will boot and load Mbed Linux.

Reviewer: ```suggestion
If you're using an Mbed Linux OS distribution it will boot and load Mbed Linux OS.
```

                        [--cert-ttl CERT_TTL] [--hab-srks]
                        [--output-dir OUTPUT_DIR]

optional arguments:
  -h, --help            show this help message and exit
  --cot-keys            Generate a TF-A chain of trust keychain, store the
                        private keys.
  --ca-ttl CA_TTL       Issuer TTL.
  --cert-ttl CERT_TTL   End entity certificate TTL.
  --hab-srks            Generate super root public keys for NXP HAB.
  --output-dir OUTPUT_DIR
                        The output directory for the HAB SRKs (public part).
```
```
usage: signctl sign-tfa [-h] [--rpi-armstub8 RPI_ARMSTUB8] --fip FIP [FIP ...]
                        [--patch-rotkey] [--output-dir OUTPUT_DIR]

optional arguments:
  -h, --help            show this help message and exit
  --rpi-armstub8 RPI_ARMSTUB8
                        Path to the armstub8.bin
  --fip FIP [FIP ...]   Path to one or more FIPs to be signed.
  --patch-rotkey        Patch root of trust public key hash in to BL1 and BL2.
  --output-dir OUTPUT_DIR
                        Output dir for signed images and the original fip
                        components.
```

## Generating Chain of Trust (CoT) Keychain

> You can generate the keychain used for the TF-AS CoT using the `signctl generate --cot-keys` command.

Reviewer: ```suggestion
You can generate the keychain used for the TF-A CoT using the `signctl generate --cot-keys` command.
```
